capsules/src/usb/descriptors.rs

```diff
@@ -370,6 +375,183 @@ impl Descriptor for DeviceDescriptor {
     }
 }
 
+/// Buffer for holding the device descriptor.
+// TODO it's dumb that these are Cells, but doing otherwise would require
+// rewriting the `write_to` functions
+pub struct DeviceBuffer {
+    pub buf: [Cell<u8>; 19],
+    pub len: usize,
+}
+
+impl DeviceBuffer {
+    pub fn write_to(&self, buf: &[Cell<u8>]) -> usize {
+        for i in 0..self.len {
+            buf[i].set(self.buf[i].get());
+        }
+        self.len
+    }
+}
+
+/// Buffer for holding the configuration, interface(s), and endpoint(s)
+/// descriptors. Also includes class-specific functional descriptors.
+pub struct DescriptorBuffer {
+    pub buf: [Cell<u8>; 128],
+    pub len: usize,
+}
+
+impl DescriptorBuffer {
+    pub fn write_to(&self, buf: &[Cell<u8>]) -> usize {
+        for i in 0..self.len {
+            buf[i].set(self.buf[i].get());
+        }
+        self.len
+    }
+}
+
+/// Transform descriptor structs into descriptor buffers that can be
+/// passed into the control endpoint handler. Each endpoint descriptor list
+/// corresponds to the matching index in the interface descriptor list. For
+/// example, if the interface descriptor list contains `[ID1, ID2, ID3]`,
+/// and the endpoint descriptors list is `[[ED1, ED2], [ED3, ED4, ED5],
+/// [ED6]]`, then the third interface descriptor (`ID3`) has one
+/// corresponding endpoint descriptor (`ED6`).
+pub fn create_descriptor_buffers(
+    device_descriptor: DeviceDescriptor,
+    mut configuration_descriptor: ConfigurationDescriptor,
+    interface_descriptor: &mut [InterfaceDescriptor],
+    endpoint_descriptors: &[&[EndpointDescriptor]],
```

Comment:
I find it weird to separate the arrays for interface and endpoint descriptors in this way. The assumption that isn't mentioned in this API is that `interface_descriptor.len() == endpoint_descriptors.len()`.

Could it be possible to have some `InterfaceHierarchy` struct:

```rust
struct InterfaceHierarchy {
    interface: InterfaceDescriptor,
    endpoints: &[EndpointDescriptor],
}
```

Then, this function would take an `interface_hierarchies: &mut [InterfaceHierarchy]` parameter which would logically tie the interfaces to their endpoints.

Reply:
Yes, of course, but USB is a WIP stack, so I don't think we should block this change (with a tracking issue dating back years) on the most preferred design, instead we should continue our policy of forward progress is good and iterate quickly.

Reply:
Fair point. My point was mostly that in my experience working with USB, a lot of things can be brittle and therefore overlooking design could fire back in terms of long hours of subtle debugging. But sure this can be improved in a separate pull request.